ts/src/base/functions/throttle.ts

```diff
@@ -44,13 +65,48 @@ class Throttler {
         }
     }
 
+    async rollingWindowLoop() {
+        while (this.running) {
+            const { resolver, cost } = this.queue[0];
+            const nowTime = now ();
+            this.timestamps = this.timestamps.filter (t => nowTime - t.timestamp < this.config.windowSize);     // Remove timestamps outside the rolling window
+            const totalCost = this.timestamps.reduce ((sum, t) => sum + t.cost, 0);     // Calculate the total cost of requests still in the window
```

Comment:
@caoilainnl / @carlosmiei  maybe we'd better replace `.filter` and `.reduce` (as they are not good performance-wise) with manual loops (`for/while`) which should take less computation.